### DISCUSSION TITLE
Fix the delegate type system

### DISCUSSION THREAD

**Discussion by Banyc** [dotnet/csharplang#4536] [Language Ideas]
## Background and Motivation

I want to clean up all the delegates that are previously assigned to a timer's `Elapsed` event. I don't want it to fire the old delegates that have had the old parameters. I want the timer to fire the newest delegate with the newest parameters.

From this small perspective, we can find out that there is a pitfall in the delegate type system. It might be a C# language design flaw. Based on the contributor of dotnet/runtime, it is about the ability to modify the contract of delegate and event. I hope the contributors of both repositories can have a discussion of this issue since I, as a common user, have not dived into the .NET platform.

## Proposed API

Please make EventHandler as a IList that could be Clear(), Add(), ...

Just make something like:

```csharp
public class EventHandler : IList<Delegate>
```

or

```csharp
public class EventHandler : List<Delegate>
```

## Usage Examples

```csharp
timer.Elapsed.Clear();
```

## Alternative Designs

There are a lot of complaints about this issue. One of the examples:

- <https://stackoverflow.com/questions/91778/how-to-remove-all-event-handlers-from-an-event>

It is not easy to remember/keep track of the old delegates that have been assigned to an event handler. Make `Clear()` method could make life tons' better.

As contributor stated [here](https://github.com/dotnet/csharplang/issues/4535), it could be solved by adding `clear` accessor to `EventHandler`. But the contributor [here](https://github.com/dotnet/runtime/issues/49595) believes the ability to modify the contract of delegate and event is the key to fix similar issues once and for all. I believe a more inclusive solution can address more problems in the future.

## Risks

Might take some time to implement and test, or even refactor all the runtime library. (I am not quite sure about that since I am just a usual user).

## Relative issues

- <https://github.com/dotnet/csharplang/issues/4535>
- <https://github.com/dotnet/runtime/issues/49595>

**Comment by CyrusNajmabadi** [comment]
I don't believe this is something related to the language. You might want to take this to dotnet/runtime. Would you like me to move it there?

**Reply by Banyc** [reply]
I don't know... Thanks for the advice and the fast reply! It might be a good idea!

**Reply by huoyaoyuan** [reply]
Updating the type system requires big work in both runtime and language.

**Comment by huoyaoyuan** [comment]
In fact, I did have some fragments of how to update the delegate type system, but didn't arrange them. In brief, my opinion is to drop multicast completely, and force multicast usage to use `List<delegate>`.

**Comment by HaloFour** [comment]
As mentioned in #4535, delegates are immutable.  When you add or remove a delegate from the event you're actually replacing the event's backing field with a new delegate instance.  There is already a means in the language to "clear" that delegate field so that it points to no subscribers: set it to `null`:

```cs
protected void ClearMyEvent() {
    this.MyEvent = null;
}
```

See here: [Example](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQEsB2MIIA4D2ANrBACarIwJYDmANDCAnJvWSANQA+AAgEwBGALAAoMbwDMAAgHSAKihjSA3mOkbZMiADcI2aQFE92ABJRMZIrmkBZAJ7H9MANzrN7jVNkAWaQCUodGQIJ2wACgBKVU9NDQcwmAB+ADoASUwdAgBrCHCYAAtg+mkyCDAoOCIYSLdROIBfWNjvXj8AYWsoBESomPq4zQSTZQBeaUwqojrGsSbxUQEZNQGvfn5+wc0dbukCYAAraXHMCAB3BSUoma3pWLj9g5Th52lOcfCQy1wS7tpkaKjAB8skEAE5wgAiADKcGAyAAxjRgDZBJDavdNI9no4Rm8Pl8yghfgh/oCQbxwVDYfCkegUQhpPx0TcttiXgZ3tJPvoiSSyccKVSYXDEcibJIWZiNNK9ocUoFgqERtdZbLsZ1oD0VRjVoN1fLFSFerrZqIGkA===)

**Reply by canton7** [reply]
(Just a note for the unwary: with delegates being immutable, delegate fields are subject to read-modify-write races. `events` [add a thread-safe wrapper around a delegate field](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYYCmAbqQHYAu6AopbQBICGVwANqQE7oBiAe0EBuZAF8gA==), but naively wrapping a field with an event having `add` and `remove` methods will not. Add your own locking if you need)

**Reply by HaloFour** [reply]
Agreed, now that C# supports `Delegate` as a generic type constraint it would be nice if the BCL would add helper methods to safely add or remove a delegate:

```cs
public class Delegate {
    public static T? Combine<T>(ref T? field, T? instance) where T : System.Delegate { ... }
    public static T? Remove<T>(ref T? field, T? instance) where T : System.Delegate { ... }
}
```

```cs
public class C {
    private EventHandler _myEvent;
    public event EventHandler MyEvent {
        add { Delegate.Combine(ref _myEvent, value); }
        remove { Delegate.Remove(ref _myEvent, value); }
    }
}
```

**Reply by CyrusNajmabadi** [reply]
Great idea. Def raise that with runtime!

**Reply by HaloFour** [reply]
@CyrusNajmabadi 

> Great idea. Def raise that with runtime!

Done and done: https://github.com/dotnet/runtime/issues/49619

**Reply by qrli** [reply]
Why having to implement explicitly? Why not simply:
```cs
public class C {
    [ThreadSafe] // by attribute or whatever means
    public event EventHandler MyEvent; // compiler generates safe field access for me
}
```

**Reply by huoyaoyuan** [reply]
The generated access should be always thread safe from the start.

**Reply by canton7** [reply]
I might have been unclear.

This is thread-safe:

```cs
public event EventHandler foo;
```

This is not thread-safe:

```cs
private EventHandler _foo;
public event EventHandler Foo
{
    add => _foo += value;
    remove => _foo -= value;
}
```
